### PR TITLE
[4.1] Tags module limit

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #36016.

### Summary of Changes
As title says. For both tag modules. 

### Testing Instructions
Use more than 20 tags to yor articles, 
Set the maximum to a number > 20 and make sure that the tags are displayed corrctly in frontend

### Actual result BEFORE applying this Pull Request
see #36016

### Expected result AFTER applying this Pull Request
Any number is possible

### CODE REVIEW

**Comment by HLeithner** [comment]
if you do it this way you have to modify the php code to allow "0" as no limit option.

**Comment by Shubhamverma2796** [comment]
Before: 
screen shot 2021-11-12 at 14 24 55

Clearly visible that the maximum count is 20

After:

screen shot 2021-11-12 at 14 28 02

Now user can set the value of the maximum count.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36019">issues.joomla.org/tracker/joomla-cms/36019</a>.</sub>

**Comment by Shubhamverma2796** [comment]
I have tested this item :white_check_mark: successfully on 38eb9e9e927a202c9b8aac40a4f547e40cf2ccd8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36019">issues.joomla.org/tracker/joomla-cms/36019</a>.</sub>

**Comment by BertaOctech** [comment]
I have tested this item :white_check_mark: successfully on 38eb9e9e927a202c9b8aac40a4f547e40cf2ccd8<br /><br />I followed the test instructions and the patch worked as described.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36019">issues.joomla.org/tracker/joomla-cms/36019</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36019">issues.joomla.org/tracker/joomla-cms/36019</a>.</sub>

**Comment by brianteeman** [comment]
@HLeithner requested valid changes. please remove RTC

**Comment by alikon** [comment]
oops missed that

**Comment by chmst** [comment]
@HLeithner  default is 5 as in other modules. The value 0 is not really useful, of course, as the result then would be empty, but for mysql this is not a  problem. 
If I change it here ( 0 = all), we must change it in all other modules too.

**Comment by HLeithner** [comment]
> @HLeithner default is 5 as in other modules. The value 0 is not really useful, of course, as the result then would be empty, but for mysql this is not a problem. If I change it here ( 0 = all), we must change it in all other modules too.

That's the reason I requested the implementation for this case in php (check for 0 and remove the setlimit call)

**Comment by chmst** [comment]
Made the changes as requested by @HLeithner . Only if the input is > 0, the maximum number is set as limit. Could you please re-test? @BertaOctech  @Shubhamverma2796 ? Thanks!

**Comment by richard67** [comment]
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36019">issues.joomla.org/tracker/joomla-cms/36019</a>.</sub>

File: modules/mod_tags_popular/src/Helper/TagsPopularHelper.php [hunk 1]
```diff
@@ -158,7 +157,11 @@ public static function getList(&$params)
 			}
 		}
 
-		$query->setLimit($maximum, 0);
+		if ($params->get('maximum','5') > 0)
```
Comment (HLeithner) [inline modules/mod_tags_popular/src/Helper/TagsPopularHelper.php]:
```suggestion
		if ((int) $params->get('maximum', 5) > 0)
```

File: modules/mod_tags_similar/src/Helper/TagsSimilarHelper.php [hunk 2]
```diff
@@ -194,7 +194,11 @@ public static function getList(&$params)
 			$query->order($query->rand());
 		}
 
-		$query->setLimit((int) $params->get('maximum', 5));
+		if ($params->get('maximum','5') > 0)
```
Comment (HLeithner) [inline modules/mod_tags_similar/src/Helper/TagsSimilarHelper.php]:
```suggestion
		if ((int) $params->get('maximum', 5) > 0)
```

**Review by HLeithner — CHANGES_REQUESTED** [review]

File: modules/mod_tags_similar/src/Helper/TagsSimilarHelper.php (hunk 2 above)
Comment (richard67) [inline modules/mod_tags_similar/src/Helper/TagsSimilarHelper.php:197]:
```suggestion
		if ($params->get('maximum', '5') > 0)
```
Fix PHP CS for drone.

File: modules/mod_tags_popular/src/Helper/TagsPopularHelper.php (hunk 1 above)
Comment (richard67) [reply modules/mod_tags_popular/src/Helper/TagsPopularHelper.php]: Why not use $maximum here? It is set at the beginning here https://github.com/joomla/joomla-cms/blob/14166af1d561ef33887d0ce6aebea01b854b52e9/modules/mod_tags_popular/src/Helper/TagsPopularHelper.php#L41 and now unused after these changes.

We could change that line 41 to `$maximum = (int) $params->get('maximum', 5);`, i.e. add the type cast to integer, and then here and below use `if ($maximum > 0)`, `$query->setLimit($maximum);` and so on.

Comment (richard67) [reply modules/mod_tags_popular/src/Helper/TagsPopularHelper.php]: Or do as suggested by Harald and remove line 41.

Comment (HLeithner) [reply modules/mod_tags_popular/src/Helper/TagsPopularHelper.php]: of course use $maximum from line 41 (cast it to int first please) and for get this suggestion

**Comment by Shubhamverma2796** [comment]
@chmst I applied the path and here are the results:

Screenshot (272)

As you can clearly see the maximum tag value can be set as negative too.  I think you should set the lower limit to  0 because obviously there is no point in setting -ve values.

File: modules/mod_tags_popular/mod_tags_popular.xml [hunk 3]
```diff
@@ -35,13 +35,10 @@
 
 				<field
 					name="maximum"
-					type="integer"
+					type="number"
 					label="MOD_TAGS_POPULAR_MAX_LABEL"
 					default="5"
 					filter="integer"
```
Comment (HLeithner) [inline modules/mod_tags_popular/mod_tags_popular.xml:41]:
```suggestion
					filter="integer"
					min="0"
```

File: modules/mod_tags_similar/mod_tags_similar.xml [hunk 4]
```diff
@@ -25,13 +25,10 @@
 			<fieldset name="basic">
 				<field
 					name="maximum"
-					type="integer"
+					type="number"
 					label="MOD_TAGS_SIMILAR_MAX_LABEL"
 					default="5"
 					filter="integer"
```
Comment (HLeithner) [inline modules/mod_tags_similar/mod_tags_similar.xml:31]:
```suggestion
					filter="integer"
					min="0"
```

**Review by HLeithner — CHANGES_REQUESTED** [review]
Add a Minimum auf 0 items

File: modules/mod_tags_popular/src/Helper/TagsPopularHelper.php (hunk 1 above)
Comment (richard67) [reply modules/mod_tags_popular/src/Helper/TagsPopularHelper.php]: @chmst This one here is not done yet.

Comment (chmst) [reply modules/mod_tags_popular/src/Helper/TagsPopularHelper.php]: done

**Comment by chmst** [comment]
The min attribute in the number field is not validated. Also with min=0 it is possible to enter every negative value. This is not in scope of this PR.

File: modules/mod_tags_popular/src/Helper/TagsPopularHelper.php [hunk 5]
```diff
@@ -158,7 +157,11 @@ public static function getList(&$params)
 			}
 		}
 
-		$query->setLimit($maximum, 0);
+		if ($maximum > 0)
+		{
+			$query->setLimit((int) $params->get('maximum'));
```
Comment (richard67) [inline modules/mod_tags_popular/src/Helper/TagsPopularHelper.php:162]:
```suggestion
			$query->setLimit($maximum);
```

**Comment by richard67** [comment]
@chmst Why closed?

**Comment by chmst** [comment]
@Shubhamverma2796 would you like to open an issue for the bug you found  in joomla? https://github.com/joomla/joomla-cms/pull/36019#issuecomment-968291979
Maybe you also find a solution ;)

**Comment by Shubhamverma2796** [comment]
@chmst  How can I open an issue because this PR is closed before getting merged, and the hard limit is still 20. Please correct me if I'm wrong.

**Comment by chmst** [comment]
The hard limit is independent from this. You found out, that it is possible to enter a negative value into a field of type="number" , also when the attribute min="0" is set. The field does not validate the input. 

You can open a new issue here 
image

and write a bug report. If you want ;)

**Comment by Shubhamverma2796** [comment]
https://user-images.githubusercontent.com/54266411/142622699-797fae8a-e78c-41d2-9019-63fd18c862c2.mp4

@chmst  In the above video, you can see that I've fixed the issue; whenever the user tries to input any -ve value, the value automatically gets changed to 0. In this way, we can easily prevent the user from entering any negative value. 
The problem is I am new here and not well-versed with the PHP and code structure of Joomla. Therefore, I cannot figure out where(in which file or folder) I should put the function written by me so that everything works fine. If anyone can guide me, then it would be my pleasure. 
Thanks 

[Link to the function](https://jsfiddle.net/4qju780h/)

**Comment by brianteeman** [comment]
@Shubhamverma2796 javascript is not the solution

**Comment by brianteeman** [comment]
> The min attribute in the number field is not validated. Also with min=0 it is possible to enter every negative value. This is not in scope of this PR.

Just add validate="number" and this will ensure that the min=0 is validated

**Comment by Shubhamverma2796** [comment]
@chmst  should I open a new pr for #36016 or will you re-open this pr itself?

**Pull request closed without merging**